File: practice/cpp/hash_table/minimum_window_substring.cpp
/*
- Problem: https://leetcode.com/problems/minimum-window-substring
*/
#include <iostream>
#include <unordered_map>
#include <string>
#include <climits>

typedef std::unordered_map<char, int> uci;

bool checkSub(uci& m_s, uci& m_t) {
  for (auto [c, num] : m_t) {
    if (m_s[c] < num) {
      return false;
    }
  }        
  return true;
}

std::string minWindow(std::string s, std::string t) {
  if (t.length() > s.length()) {
    return "";
  }
  uci m_t{}, m_s{};
  for (auto c : t) {
    m_t[c]++;
  }
        
  std::string res {};
  int count = INT_MAX;
  int pivot = 0;
  bool found = false;
  int l = 0, r = 0;
  while (r < s.length()) {
    m_s[s[r]]++;
    while (checkSub(m_s, m_t)) {
      if (r - l + 1 < count) {
        count = r - l + 1;
        pivot = l; 
      }
      m_s[s[l]]--;
      l++;
    }
    r++;
  }
        
        
  return count == INT_MAX ? "" : s.substr(pivot, count);;
}

int main(void) {
  std::cout << minWindow("ADOBECODEBANC", "ABC") << std::endl;
}

File: practice/python/hash_table/minimum_window_substring.py
"""
- Problem: https://leetcode.com/problems/minimum-window-substring
"""

def minWindow(s, t):
    def checkSub(dict_t, dict_s):
        for key in dict_t:
            if key not in dict_s.keys() or dict_t[key] > dict_s[key]:
                return False
        return True
    
    dict_t = {k: 0 for k in t}
    dict_s = {k: 0 for k in s}
    
    len_t = len(t)
    len_s = len(s)
    if len_t > len_s:
        return ""
    
    for c  in t:
        dict_t[c] += 1
    
    count = len_s + 1
    pivot = 0
    l = 0
    r = 0
    
    while r < len_s:
        dict_s[s[r]] += 1
        while True == checkSub(dict_t, dict_s):
            if r - l + 1 < count:
                pivot = l
                count = r - l + 1
            dict_s[s[l]] -= 1
            l += 1
        r += 1
        
        
    return "" if count == len_s + 1 else s[pivot : pivot + count]
        

if __name__ == "__main__":
    print(minWindow("ADOBECODEBANC","ABC"))
